execution/gethexec/executionengine.go

```diff
@@ -378,109 +389,85 @@ func (s *ExecutionEngine) HeadMessageIndexSync(t *testing.T) (arbutil.MessageInd
 }
 
-// The caller must hold the createBlocksMutex
-func (s *ExecutionEngine) resequenceReorgedMessages(messages []*arbostypes.MessageWithMetadata) {
-	if !s.reorgSequencing {
-		return
+func (s *ExecutionEngine) blockCreationStopped() bool {
+	return s.syncTillBlock > 0 && s.latestBlock != nil && s.latestBlock.NumberU64() >= s.syncTillBlock
+}
+
+func (s *ExecutionEngine) ResequenceReorgedMessage(msg *arbostypes.MessageWithMetadata) (*execution.SequencedMsg, error) {
+	if s.blockCreationStopped() {
+		return nil, ExecutionEngineBlockCreationStopped
 	}
+	s.createBlocksMutex.Lock()
+	defer s.createBlocksMutex.Unlock()
 
-	log.Info("Trying to resequence messages", "number", len(messages))
 	lastBlockHeader, err := s.getCurrentHeader()
 	if err != nil {
-		log.Error("block header not found during resequence", "err", err)
-		return
+		return nil, fmt.Errorf("failed to get current block header: %w", err)
 	}
 
 	nextDelayedMsgIdx := lastBlockHeader.Nonce.Uint64()
 
-	for _, msg := range messages {
-		// Check if the message is non-nil just to be safe
-		if msg == nil || msg.Message == nil || msg.Message.Header == nil {
-			continue
-		}
-		header := msg.Message.Header
-		if header.RequestId != nil {
-			delayedMsgIdx := header.RequestId.Big().Uint64()
-			if delayedMsgIdx != nextDelayedMsgIdx {
-				log.Info("not resequencing delayed message due to unexpected index", "expected", nextDelayedMsgIdx, "found", delayedMsgIdx)
-				continue
-			}
-			_, err := s.sequenceDelayedMessageWithBlockMutex(msg.Message, delayedMsgIdx)
-			if err != nil {
-				log.Error("failed to re-sequence old delayed message removed by reorg", "err", err)
-			}
-			nextDelayedMsgIdx += 1
-			continue
-		}
-		if header.Kind != arbostypes.L1MessageType_L2Message || header.Poster != l1pricing.BatchPosterAddress {
-			// This shouldn't exist?
-			log.Warn("skipping non-standard sequencer message found from reorg", "header", header)
-			continue
-		}
-		lastArbosVersion := types.DeserializeHeaderExtraInformation(lastBlockHeader).ArbOSFormatVersion
-		txes, err := arbos.ParseL2Transactions(msg.Message, s.bc.Config().ChainID, lastArbosVersion)
-		if err != nil {
-			log.Warn("failed to parse sequencer message found from reorg", "err", err)
-			continue
+	// Check if the message is non-nil just to be safe
+	if msg == nil || msg.Message == nil || msg.Message.Header == nil {
+		return nil, nil
+	}
+	header := msg.Message.Header
+	if header.RequestId != nil {
+		delayedMsgIdx := header.RequestId.Big().Uint64()
+		if delayedMsgIdx != nextDelayedMsgIdx {
+			log.Info("not resequencing delayed message due to unexpected index", "expected", nextDelayedMsgIdx, "found", delayedMsgIdx)
+			return nil, nil
 		}
-		hooks := MakeZeroTxSizeSequencingHooksForTesting(txes, nil, nil, nil)
-		block, err := s.sequenceTransactionsWithBlockMutex(msg.Message.Header, hooks, nil)
+		sequencedMsg, err := s.sequenceDelayedMessageWithBlockMutex(msg.Message, delayedMsgIdx)
```

Comment:
In `SequenceDelayedMessage()` we acquire both locks `delayedMsgsMutex` and `createBlocksMutex` before calling `sequenceDelayedMessageWithBlockMutex()`, while here we just acquire `createBlocksMutex`. Wouldn't we also need to acquire `delayedMsgsMutex` here? I thought maybe the caller of `ResequenceReorgedMessage` would acquire such lock but couldn't find there either.

Reply:
delayedMsgsMutex controls access to the delayedMsgs queue.
This procedure doesn't access it, so no need to lock delayedMsgsMutex.